synedrion/src/cggmp21/key_refresh_tests.rs

```diff
@@ -0,0 +1,688 @@
+use alloc::collections::BTreeSet;
+
+use manul::{
+    combinators::misbehave::Misbehaving,
+    dev::{BinaryFormat, TestSessionParams, TestSigner, TestVerifier},
+    protocol::{
+        Artifact, BoxedRound, Deserializer, DirectMessage, EchoBroadcast, EntryPoint, LocalError, NormalBroadcast,
+        ProtocolMessagePart, Serializer,
+    },
+    signature::Keypair,
+};
+use rand_chacha::ChaCha8Rng;
+use rand_core::{CryptoRngCore, OsRng, SeedableRng};
+
+use super::{
+    key_refresh::{
+        KeyRefresh, Round1, Round1EchoBroadcast, Round2Broadcast, Round2EchoBroadcast, Round3Broadcast,
+        Round3DirectMessage, Round3EchoBroadcast,
+    },
+    params::{SchemeParams, TestParams},
+    sigma::{FacProof, ModProof, PrmProof, SchCommitment, SchProof, SchSecret},
+};
+use crate::{
+    curve::Scalar,
+    paillier::{PaillierParams, PublicKeyPaillierWire, RPParams, RPParamsWire, RPSecret, SecretKeyPaillierWire},
+    tools::{
+        hashing::FofHasher,
+        protocol_shortcuts_dev::{check_evidence_with_behavior, check_invalid_message_evidence, CheckPart},
+        Secret,
+    },
+};
+
+type Id = TestVerifier;
+type P = TestParams;
+type SP = TestSessionParams<BinaryFormat>;
+
+#[allow(clippy::type_complexity)]
+fn make_entry_points() -> (BTreeSet<Id>, Vec<(TestSigner, KeyRefresh<P, Id>)>) {
+    let signers = (0..3).map(TestSigner::new).collect::<Vec<_>>();
+    let all_ids = signers
+        .iter()
+        .map(|signer| signer.verifying_key())
+        .collect::<BTreeSet<_>>();
+
+    let entry_points = signers
+        .into_iter()
+        .map(|signer| (signer, KeyRefresh::new(all_ids.clone()).unwrap()))
+        .collect();
+    (all_ids, entry_points)
+}
+
+fn check_evidence<M>(expected_description: &str) -> Result<(), LocalError>
+where
+    M: Misbehaving<Id, (), EntryPoint = KeyRefresh<P, Id>>,
+{
+    let (associated_data, entry_points) = make_entry_points();
+    check_evidence_with_behavior::<SP, M, _>(&mut OsRng, entry_points, &(), &associated_data, expected_description)
+}
+
+#[test]
+fn invalid_messages() {
```

Comment:
Echo broadcast tests to be added after https://github.com/entropyxyz/manul/pull/86 is in